akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -51,12 +51,15 @@ private[http] object HttpServerBluePrint {
     stack(settings, remoteAddress, log)
 
   def stack(settings: ServerSettings, remoteAddress: Option[InetSocketAddress], log: LoggingAdapter)(implicit mat: Materializer): Http.ServerLayer = {
-    userHandlerGuard(settings.pipeliningLimit) atop
-      requestPreparation(settings, remoteAddress) atop
-      controller(settings, log) atop
-      parsingRendering(settings, log) atop
-      websocketSupport(settings, log) atop
-      unwrapTls
+    val theStack =
+      userHandlerGuard(settings.pipeliningLimit) atop
+        requestPreparation(settings) atop
```

Comment:
Well, to at least show the other evil, here it is: https://github.com/ktoso/akka/commit/98a068ab12552c253ddbc82610eb2f725b8ef44a (Source.maybe + completing it with the remote address)

It looks and feels very weird and it's hard to track down all the places where one has to complete the maybe source to get the server running. Technically `serverLayer` is even user-api AFAIR, so we'd need to document "yeah and then you need to get the materialized value and complete it" which sounds _really weird_.

It of course also is slower, since we're halting any processing until stream is materialized -> promise is completed -> future completes, triggering the zip. We _have_ the value, and setting it early, as attribute, makes a lot of sense from a usability, maintenance as well as performance perspective. 

Thus, I'm suggesting to stick with the `attribute + graphStage` as implemented in https://github.com/ktoso/akka/commit/99e40f918b482a89f56ff5c789d0a75ef7f09563 and one question which should be answered then is: do we make accessing that attribute an actual user API (as @jrudolph suggested), or do we hide it (needs `InternalAttribute` marker interfaces, but is doable ofc).